File: iosgl/iosgl/bt/display/particle/ctrl/Curve.m
//
//  Curve.m
//  iosgl
//
//  Created by zhao on 14/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Curve.h"
#import "CurveVo.h"
#import "GL_Header.h"
#import "Vector3D.h"
#import "Scene_data.h"

@implementation Curve

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.valueV3d=[[[NSArray alloc]initWithObjects:@1,@1,@1,@1, nil] mutableCopy];
    }
    return self;
}


-(void)setData:(CurveVo*)obj;
{
    Curve* this=self;
    
    this.type=obj.type;
    this.maxFrame = obj.maxFrame;
    if (obj.items.count) {
        this.begintFrame = [[obj.items[0] objectForKey:@"frame"]intValue] ;
    } else {
        this.begintFrame = -1;
    }
    NSArray* values0  = obj.values[0];
    NSUInteger len = values0.count;
    NSMutableArray<NSMutableArray*>* ary=[[NSMutableArray alloc]init];
    for (int i=0; i < len; i++) {
        NSArray<NSNumber*>* values0  = obj.values[0];
        NSArray<NSNumber*>* values1  = obj.values[1];
        NSArray<NSNumber*>* values2  = obj.values[2];
        NSArray<NSNumber*>* values3  = obj.values[3];
        NSMutableArray* itemAry  = [[NSMutableArray alloc]init];
        switch (this.type) {
            case 1:
                [itemAry addObject:values0[i]];
                break;
            case 2:
                [itemAry addObject:values0[i]];
                [itemAry addObject:values1[i]];
                break;
            case 3:
                [itemAry addObject:values0[i]];
                [itemAry addObject:values1[i]];
                [itemAry addObject:values2[i]];
                break;
            case 4:
                [itemAry addObject: [NSNumber numberWithFloat:values3[i].floatValue*values0[i].floatValue]];
                [itemAry addObject: [NSNumber numberWithFloat:values3[i].floatValue*values1[i].floatValue]];
                [itemAry addObject: [NSNumber numberWithFloat:values3[i].floatValue*values2[i].floatValue]];
                [itemAry addObject: values3[i]];
                break;
            default:
                break;
        }
        [ary addObject:itemAry];
    }
    this.valueVec = ary;
}
-(NSMutableArray<NSNumber*>*)getValue:(float)t;
{
    Curve* this=self;
    if (!this.valueVec || this.begintFrame == -1) {
        return this.valueV3d;
    }
    int flag =  float2int(t / [Scene_data default].frameTime- this.begintFrame);
    if (flag < 0) {
        flag = 0;
    } else if (flag > this.maxFrame - this.begintFrame) {
        flag = this.maxFrame - this.begintFrame;
    }
    NSLog(@"flag %d    %f   %f   %d",flag,t , [Scene_data default].frameTime, this.begintFrame);
    return this.valueVec[flag];;
}
/*
public getValue($t: number): Array<number> {
       if (!this.valueVec || this.begintFrame == -1) {
           return this.valueV3d;
       }
       var flag: number = float2int($t / Scene_data.frameTime - this.begintFrame);

       if (flag < 0) {
           flag = 0;
       } else if (flag > this.maxFrame - this.begintFrame) {
           flag = this.maxFrame - this.begintFrame;
       }

       return this.valueVec[flag];
*/
@end

File: baseSrc/android/app/src/main/java/z3d/material/Curve.java
package z3d.material;

import java.util.ArrayList;
import java.util.List;
import java.util.Vector;

import z3d.base.Scene_data;
import z3d.vo.CurveItemVo;
import z3d.vo.CurveVo;

public class Curve {

    public  int type;
    public  int begintFrame;
    public  int maxFrame;
    public List<List<Float>> valueVec;
    public List<Float> valueV3d;

    public  Curve(){
        valueV3d=new ArrayList<>();
        valueV3d.add(1.0f);
        valueV3d.add(1.0f);
        valueV3d.add(1.0f);
        valueV3d.add(1.0f);

    }
    public List<Float> getValue(float t) {
        if (this.valueVec==null || this.begintFrame == -1) {
            return this.valueV3d;
        }
        int flag =  float2int(t / Scene_data.frameTime - this.begintFrame);
        if (flag < 0) {
            flag = 0;
        } else if (flag > this.maxFrame - this.begintFrame) {
            flag = this.maxFrame - this.begintFrame;
        }
        return this.valueVec.get(flag);
    }
    private  int float2int(float val){

        return  (int )val;
    }

    public void setData(CurveVo obj) {

        this.type = obj.type;
        this.maxFrame = obj.maxFrame;
        if (obj.items.size()>0) {
            CurveItemVo curveItemVo=(CurveItemVo) obj.items.get(0);
            this.begintFrame = curveItemVo.frame;
        } else {
            this.begintFrame = -1;
        }
        int len = obj.values.get(0).size();
        List<List<Float>> ary= new ArrayList<>();
        for (int i = 0; i < len; i++) {
            List<Float> itemAry  = new ArrayList<>();
            if (this.type == 1) {
                itemAry.add(obj.values.get(0).get(i));
            } else if (this.type == 2) {
                itemAry.add(obj.values.get(0).get(i));
                itemAry.add(obj.values.get(1).get(i));
            } else if (this.type == 3) {
                itemAry.add(obj.values.get(0).get(i));
                itemAry.add(obj.values.get(1).get(i));
                itemAry.add(obj.values.get(2).get(i));
            } else if (this.type == 4) {
                Float w = obj.values.get(3).get(i);
                itemAry.add(obj.values.get(0).get(i)*w);
                itemAry.add(obj.values.get(1).get(i)*w);
                itemAry.add(obj.values.get(2).get(i)*w);
                itemAry.add(w);
            }
            ary.add(itemAry);
        }
        this.valueVec = ary;


    }
}
